### ISSUE TITLE
TileRDDReproject Produces noDataTiles in the Resulting Layer

### ISSUE THREAD

**Issue by jbouffard** [locationtech/geotrellis#2798]
## Overview

`TileRDDReproject` produces `noDataTile`s in the layer that it creates. This behavior was discovered during the testing of `RasterSourceRDD`, [an object](https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/main/scala/RasterSourceRDD.scala) from the [geotrellis-contrib repository](https://github.com/geotrellis/geotrellis-contrib). Therefore, the following examples will use code from that repo with explanations provided.

### Background

`RasterSourceRDD` is an object that is able to produce a `TileLayerRDD[K]` from some [RasterSource](https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/main/scala/RasterSource.scala) and a `LayoutDefinition`. A `RasterSource` being some data source that can be represented as a Raster. In this example, we will be working with [GeoTiffRasterSource](https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/main/scala/GeoTiffRasterSource.scala), a data source that is from `GeoTiff`s. Thus, we are taking an arbitrary `GeoTiff` and turning it into a `TileLayerRDD[K]` given some layout.

In addition to producing a `TileLayerRDD[K]`, it's also possible to perform reprojection during that step as well. This means then that reading, reprojection, and tiling all occur at the same time. Which helps improve performance. In order to perform this reprojection, a [WarpGeoTiffRasterSource](https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/main/scala/WarpGeoTiffRasterSource.scala) must be used. `WarpGeoTiffRasterSource` is similar to `GeoTiffRasterSource` except that it'll produce the source data in the desired projection.

### The Tests

By using `RasterSourceRDD`, we are able to simplify the process of reading, tiling, and reprojecting unstructured raster data. Therefore, we should expect the same results between `RasterSourceRDD` and our traditional way of producing layers.

#### Expected Result

For the expected result, we use `HadoopGeoTiffRDD` which we then tile and reproject:

```scala
  val uri = s"file://${new File("").getAbsolutePath()}/src/test/resources/img/aspect-tiled.tif"
  val rasterSource = GeoTiffRasterSource(uri)

  // The desired projection for the resulting layer
  val targetCRS = CRS.fromEpsgCode(3857)

  val scheme = ZoomedLayoutScheme(targetCRS)

  // The desired layout for the resulting layer
  val layout = scheme.levelForZoom(13).layout

  val floatingLayout = FloatingLayoutScheme(256)
  val geoTiffRDD = HadoopGeoTiffRDD.spatialMultiband(uri)
  val md = geoTiffRDD.collectMetadata[SpatialKey](floatingLayout)._2

  // We first tile the RDD to a local layout and then reproject it to the desired CRS and layout
  val reprojectedExpectedRDD: MultibandTileLayerRDD[SpatialKey] = {
    geoTiffRDD
      .tileToLayout(md)
      .reproject(
        targetCRS,
        layout,
        Reproject.Options(targetCellSize = Some(layout.cellSize))
      )._2.persist()
```

#### First Test: Tiling and then Reprojection

The first test was to first create a `TileLayerRDD[K]` using `RasterSourceRDD` and then performing `reproject` on the following layer.

```scala
        // This should be the same as result of .tileToLayout(md.layout)
        val rasterSourceRDD: MultibandTileLayerRDD[SpatialKey] =
          RasterSourceRDD(rasterSource, md.layout)

        // Complete the reprojection
        val reprojectedSource =
          rasterSourceRDD.reproject(targetCRS, layout)._2

        // SpatialKeys that cover the `RasterSource`
        val keys = {
          val inter = layout.mapTransform.extent.intersection(rasterSource.withCRS(targetCRS).extent).get
          layout.mapTransform.keysForGeometry(inter.toPolygon)
        }

        // The values in the expected result that are noDataTiles
        val results = reprojectedExpectedRDD.filter { case (k, v) =>
          v.bands.map { _.isNoDataTile }.reduce { _ && _ }
        }

        // The values in the actual result that are noDataTiles
        val results2 = reprojectedSource.filter { case (k, v) =>
          v.bands.map { _.isNoDataTile }.reduce { _ && _ }
        }

        val aKeys = reprojectedSource.keys.collect().toSet
        val eKeys = reprojectedExpectedRDD.keys.collect().toSet

        println(s"\nRasterSourceRDD with a GeoTiffRasterSource")
        println(s"This is the number of keys that intersect with the extent: ${keys.size}")
        println(s"This is the number of elements in the reprojectedSourceRDD: ${reprojectedSource.count}")
        println(s"This is the number of elements in the reprojectedExpectedRDD: ${reprojectedExpectedRDD.count}")
        println(s"This is the number of noDataTiles in the the reprojectedExpectedRDD: ${results.count}")
        println(s"This is the number of noDataTiles in the the reprojectedSource: ${results2.count}")
        println(s"These are keys with the noDataTiles: ${results.keys.collect().toSet}\n")
        println(s"These are the keys that are not found in reprojectedSourceRDD: ${eKeys.diff(aKeys)}\n")

        assertRDDLayersEqual(reprojectedExpectedRDD, reprojectedSource)
```

This test past, however, some debugging `println`s show something interesting:

```
RasterSourceRDD with a GeoTiffRasterSource
This is the number of keys that intersect with the extent: 20
This is the number of elements in the reprojectedSourceRDD: 25
This is the number of elements in the reprojectedExpectedRDD: 25
This is the number of noDataTiles in the the reprojectedExpectedRDD: 5
This is the number of noDataTiles in the the reprojectedSource: 5

These are keys with the noDataTiles: Set(SpatialKey(2306,3226), SpatialKey(2307,3226), SpatialKey(2305,3226), SpatialKey(2303,3226), SpatialKey(2304,3226))

These are the keys that are not found in reprojectedSourceRDD: Set()
```

We can see that the number of keys covered by the `RasterSource` is only `20`, yet both `reprojectedSource` and `reprojectedExpectedRDD` have `25` elements. In addition, `reprojectedSource` and `reprojectedExpectedRDD` contains `5` `noDataTile`s.

#### Second Test: Tiling and Reprojection in a Single Step

This next test produced a `TileLayerRDD[K]` that was both tiled and reprojected in a single step via `WarpGeoTiffRasterSource` and `RasterSourceRDD`.

```scala
        // Reprojection is handled by the `WarpGeoTiffRasterSource` and tiling is done by
        // `RasterSourceRDD`
        val reprojectedSourceRDD: MultibandTileLayerRDD[SpatialKey] =
          RasterSourceRDD(rasterSource.withCRS(targetCRS), layout)

        // SpatialKeys that cover the `RasterSource`
        val keys = {
          val inter = layout.mapTransform.extent.intersection(rasterSource.withCRS(targetCRS).extent).get
          layout.mapTransform.keysForGeometry(inter.toPolygon)
        }

        // The values in the expected result that are noDataTiles
        val results = reprojectedExpectedRDD.filter { case (k, v) =>
          v.bands.map { _.isNoDataTile }.reduce { _ && _ }
        }

        // The values in the actual result that are noDataTiles
        val results2 = reprojectedSourceRDD.filter { case (k, v) =>
          v.bands.map { _.isNoDataTile }.reduce { _ && _ }
        }

        val aKeys = reprojectedSourceRDD.keys.collect().toSet
        val eKeys = reprojectedExpectedRDD.keys.collect().toSet

        println(s"\nRasterSourceRDD with a WarpGeoTiffRasterSource")
        println(s"This is the number of keys that intersect with the extent: ${keys.size}")
        println(s"This is the number of elements in the reprojectedSourceRDD: ${reprojectedSourceRDD.count}")
        println(s"This is the number of elements in the reprojectedExpectedRDD: ${reprojectedExpectedRDD.count}")
        println(s"This is the number of noDataTiles in the the reprojectedExpectedRDD: ${results.count}")
        println(s"This is the number of noDataTiles in the the reprojectedSourceRDD: ${results2.count}")
        println(s"These are keys with the noDataTiles: ${results.keys.collect().toSet}\n")
        println(s"These are the keys that are not found in reprojectedSourceRDD: ${eKeys.diff(aKeys)}\n")

       assertRDDLayersEqual(
         reprojectedExpectedRDD,
         reprojectedSourceRDD
       )
```

Unlike the previous test, this one fails due to the following:

```
[info]   - should reproduce tileToLayout followed by reproject *** FAILED ***
[info]     java.lang.Exception: SpatialKey(2305,3226) does not exist in the rasterSourceRDD
[info]     at geotrellis.contrib.vlm.RasterSourceRDDSpec$$anonfun$1$$anonfun$geotrellis$contrib$vlm$RasterSourceRDDSpec$$anonfun$$assertRDDLayersEqual$1$1.apply(RasterSourceRDDSpec.scala:118)
[info]     at geotrellis.contrib.vlm.RasterSourceRDDSpec$$anonfun$1$$anonfun$geotrellis$contrib$vlm$RasterSourceRDDSpec$$anonfun$$assertRDDLayersEqual$1$1.apply(RasterSourceRDDSpec.scala:115)
[info]     at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
[info]     at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
[info]     at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[info]     at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[info]     at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
[info]     at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
[info]     at geotrellis.contrib.vlm.RasterSourceRDDSpec$$anonfun$1.geotrellis$contrib$vlm$RasterSourceRDDSpec$$anonfun$$assertRDDLayersEqual$1(RasterSourceRDDSpec.scala:115)
[info]     at geotrellis.contrib.vlm.RasterSourceRDDSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$3.apply$mcV$sp(RasterSourceRDDSpec.scala:183)
```

We can see that the two resulting layers are not equal because `reprojectedExpectedRDD` contains keys that `reprojectedSourceRDD` does not. Through the `println`s, we can get a clearer understanding of why the test is failing.

```
RasterSourceRDD with a WarpGeoTiffRasterSource
This is the number of keys that intersect with the extent: 20
This is the number of elements in the reprojectedSourceRDD: 20
This is the number of elements in the reprojectedExpectedRDD: 25
This is the number of noDataTiles in the the reprojectedExpectedRDD: 5
This is the number of noDataTiles in the the reprojectedSourceRDD: 0
These are keys with the noDataTiles: Set(SpatialKey(2306,3226), SpatialKey(2307,3226), SpatialKey(2305,3226), SpatialKey(2303,3226), SpatialKey(2304,3226))

These are the keys that are not found in reprojectedSourceRDD: Set(SpatialKey(2306,3226), SpatialKey(2307,3226), SpatialKey(2305,3226), SpatialKey(2303,3226), SpatialKey(2304,3226))
```

Just like before, `reprojectedSourceRDD` only cover `20` keys in the desired layout. However, unlike earlier we can see that `reprojectedSourceRDD` contains just `20` elements, which we would expect, while `reprojectedExpectedRDD` contains `25`. `5` of them being `noDataTile`s. Looking at the keys of the `noDataTile`s, we can see that they are the same ones that are absent from `reprojectedSourceRDD`, and that `reprojectedSourceRDD` itself has `0` `noDataTile`s.

### Conclusion

The only difference between the first and second test was the intermediate `reproject` call in the first. Therefore, it appears that the `reproject` method produces a layer with `noDataTile`s. The exact cause of this issue is not known, but it doesn't seem to be causing any problems outside of this test for now. Thus, this is something we should be mindful of in our future tests.

### tl;dr

`TileRDDReproject` produces `noDataTile`s outside the `Extent` of the layer. I'm not sure why they're appearing, but they shouldn't cause too much issue outside of some tests.
